webui/react/src/components/kit/CodeEditor.tsx

```diff
@@ -31,12 +31,28 @@ const MARKDOWN_CONFIG = {
   highlightActiveLineGutter: false,
 };
 
+type ErrorMessage = {
+  _tag: 'Error';
+  message: string;
+};
+// extension to loadable
+export type LoadableOrError<T> = Loadable<T> | ErrorMessage;
+
+export const ErrorMessage = (message: string): ErrorMessage => ({
+  _tag: 'Error' as const,
+  message,
+});
+
+const isErrorMessage = (f: unknown): f is ErrorMessage =>
+  !!f && typeof f === 'object' && '_tag' in f && f._tag === 'Error';
+
 export type Props = {
   files: TreeNode[];
-  onError: ErrorHandler;
+  file: LoadableOrError<string>;
+  onError: ErrorHandler; // only used to raise ipynb errors
   height?: string; // height of the editable area, if a title is provided that will add an additional ~38px
-  onChange?: (arg0: string) => void; // only use in single-file editing
-  onSelectFile?: (arg0: string) => void;
+  onChange?: (fileContent: string) => void; // only use in single-file editing
+  onSelectFile?: (filePath: string) => void;
   readonly?: boolean;
   selectedFilePath?: string;
```

Comment:
Based on your usage in the design kit, should this be required? Also if the component is only available as controlled (which is my understanding) then `onSelectFile` should also be required, right?

Reply:
`selectedFilePath`/`onSelectFile` aren't necessary in single file applications.

Reply:
Can you account for that by having one variant of props take `files: [TreeNode]` and one take `files: TreeNode[]`?